README
Mitsuba renderer v2.0.2 (c) 2010 by Wenzel Jakob

Features:
    - Runs on Linux (32/64-bit), MacOS X (x86/PPC) and Microsoft 
      Windows (32-bit).
    - Written in portable C++ with optional SSE2 optimizations when compiled 
      for the i386 and x86_64 platforms.
    - Modular architecture inspired by PBRT.
    - Fully parallelized for multi-core operation.
    - Supports interactive walkthroughs using OpenGL and real-time coherent ray tracing.
    - Network rendering - can distribute the computational load of a single job 
      over a cluster without requiring a shared filesystem.
    - For convenience, Mitsuba can also transparently SSH into a cluster, 
      remotely launch itself and communicate through the encrypted link.
    - Uses a O(n log n) KD-tree compiler with primitive clipping support
      and Havran's fast traversal algorithm.
    - Highly optimized SSE2 intersection code for triangle meshes. Custom
      shapes such as spheres and cylinders are also supported.
    - High dynamic-range (EXR) support for textures and generated images
    - Cross-platform hardware rendering abstraction layer.
    - BSDFs: Lambertian, Perfect Mirror, Phong, Dielectric, Anisotropic Ward,
             Microfacet BRDF, Rough Glass (Walter et al.), 
    - Supports homogeneous as well as heterogeneous participating media
    - Implements Jensen's dipole approximations to sub-surface scattering
    - Supports a variety of image reconstruction filters 
      (eg. Mitchell-Netravali, Catmull-Rom, Gaussian, Lanczos-Sinc, ..) 
    - Comes with several numerical testbenches, which verify rendered 
      test scenes against analytic results.
    - Full spectral rendering is supported.
    - Light sources: area lights, point lights, environment maps, 
	  collimated beams.
    - QMC sampling
	- Cluster job submission support using PBS
    - Low dynamic-range textures (JPEG and PNG)
    - Depth of field
    - Adaptive/error-estimating sampling

External dependencies:
    A recent C++ compiler (GCC 4.1-4.3 or Visual Studio 2005/2008)
    SCons (An advanced build system built on top of Python)
    Xerces-C (For scene description XML parsing)
    GLEW (For the hardware acceleration abstraction layer)
    OpenEXR (Needed for high dynamic-range image file export/import)
	Boost (An extensive C++ class library)
	libpng & libjpeg (For PNG & JPG image import/export)

    When building Mitsuba on Windows or OSX, it is recommended to use
    the pre-compiled versions of these dependencies, which are supplied 
    with the source code. If you follow the instructions below, this
	will happen automatically.

Building:
    Go to the "config" directory and copy the "config.py" version which 
    best matches your environment. Some minor adaptions may be 
    required. The following configurations are currently provided:

    config-linux.py          - Linux build (GCC 4.2+ recommended)
    config-linux-icc.py      - Linux build using the Intel compiler
    config-darwin-x86.py     - OSX build targeting x86 using GCC4.2
    config-darwin-ppc.py     - OSX build targeting PPC using GCC4.2
    config-msvc2005-win32.py - Windows build using MSVC 2005 or 2008 (32 bit)
    config-msvc2005-win64.py - Windows build using MSVC 2005 or 2008 (64 bit)

	Make sure that you have a working install of Python. On the Windows
	platform, SCons (http://www.scons.org/) must also be installed before
	building.

    Next, run the script "tools/build.sh" (Linux, OSX) or "scons" (Windows) 
	to start compiling using the supplied version of SCons. On Windows, this 
	must be done on the Visual Studio command prompt (e.g by launching 
	Start->Programs->Microsoft Visual Studio 2005->
		Visual Studio Tools->Visual Studio 2005 Command Prompt).

	If Windows complains about a missing library named 'MSVCR80.dll',
	please install the redistributable in "tools/windows/bin". Also,
	be sure to add the path containing the Qt tools (e.g. 'moc.exe')
	to the PATH.

    To build an universal binary on OSX, enter
 
        $ tools/darwin/build-universal.sh

    and have a cup of coffee :).

Compilation flags:
    Some of the compilation flags in "config.py" have a big impact on
    the resulting executable:

    'MTS_DEBUG'           - Enable assertions etc.. Usually a good idea.
    'SINGLE_PRECISION'    - Use single precision floating point values
    'DOUBLE_PRECISION'    - Use double precision floating point values
    'MTS_SSE'             - Activate optimized SSE routines.
    'MTS_HAS_COHERENT_RT' - Include coherent ray tracing support (needs SSE)
    'MTS_DEBUG_FP'        - Generated NaNs will cause FP exceptions, which
	                        can be caught in a debugger. Warning: This is slow!

    Single precision is sufficient in most cases. The use of double precision
    currently precludes the SSE feature, since the accelerated code has only
    been written as a single-precision version.

Getting started:
    On Linux/OSX, the library path must be manually set: on the command line, 
    this can be done using the script "setpath.sh":

        $ . setpath.sh

	On Windows, all relevant files are installed to 'dist' and the binaries
	should be executed from there. 

	On Linux, an interactive preview can be launched by entering (e.g.)

        $ mtsgui scenes/cornell/box.xml

    Rendering can be started by pressing the 'r' key. 

	The 'mitsuba' binary is an alternative non-interactive rendering 
	frontend for command-line use and batch job operation.

    To get a listing of the supported command-line parameters, run
    the executables without parameters:

	$ mitsuba 

	Mitsuba version 2.0.2, Copyright (c) 2010 Wenzel Jakob
	Usage: mitsuba [options] <One or more scene XML files>
	Options/Arguments:
	-h          Display this help text

	-D key=val  Define a constant, which can referenced as "$key" in the scene

	-o fname    Write the output image to the file denoted by "fname"

	-a p1;p2;.. Add one or more entries to the resource search path

	-p count    Override the detected number of processors. Useful for reducing
				the load or creating scheduling-only nodes in conjunction with
				the -c and -s parameters, e.g. -p 0 -c host1;host2;host3,...

	-q          Quiet mode - do not print any log messages to stdout

	-c hosts    Network rendering: connect to mtssrv instances over a network.
				Requires a semicolon-separated list of host names of the form
						host.domain[:port] for a direct connection
					or
						user@host.domain[:path] for a SSH connection (where
						"path" denotes the place where Mitsuba is checked
						out -- by default, "~/mitsuba" is used)

	-s file     Connect to additional Mitsuba servers specified in a file
				with one name per line (same format as in -c)

	-j count    Simultaneously schedule several scenes. Can sometimes accelerate
				rendering when large amounts of processing power are available
				(e.g. when running Mitsuba on a cluster. Default: 1)

	-n name     Assign a node name to this instance (Default: host name)

	-t          Test case mode (see Mitsuba docs for more information)

	-v          Be more verbose

	-b          Disable progress bars

	The README file included with the distribution contains further information.

	On Windows, a scene file can simply be dropped on the 'mtsgui.exe'
	executable to start the interactive preview.

	When doing lengthy command line renders on Linux or OSX, it is possible 
	to send a signal to the process using 'kill -HUP', causing Mitsuba to 
	write out the partially finished image and continue rendering.

Network rendering:
    For extremely lengthy rendering tasks, it may make sense to spread the 
    computational load of a single job over multiple machines. Mitsuba 
    supports this via two mechanisms: The easiest and most secure method can
    only be used when Linux or OSX is installed on the server machine
    machine:

    This works as follows: The client creates an encrypted SSH connection 
	to the server, launches a Mitsuba worker instance on the remote end 
	and tunnels all subsequent communication over the encrypted link. All 
	of this can be done simply by running

        $ mitsuba -c user1@host:path-to-mitsuba-install path/to/scene.xml

    Passwordless logins must be enabled for this to work, which generally
	entails installing a suitable "~/.ssh/authorized_keys2" file on the server
	and running "ssh-agent" on the client. On Windows, the situation is a bit
	more complicated because it does not come with a suitable SSH client by
	default. To get SSH to work, Mitsuba requires plink.exe (from PUTTY) to
	be on the $PATH. For passwordless authentication with a Linux/OSX-based
	server, convert your private key to PuTTY's format (with the help 
	of puttygen.exe). Afterwards, pageant.exe is required to load and 
	authenticate the key. All of these binaries are available in
	'tools/windows/bin'

	The alternative to SSH-based communication is to run a dedicated mitsuba
    server:

        $ mtssrv
  
    By default, the server is configured to listen on port 7554. Clients can
	then connect using the following syntax:

        $ mitsuba -c host1;host2:1234 path/to/scene.xml

    Note that it is possible to use both connection mechanisms at the same
    time and connect to, say, 3 SSH-based servers and 10 dedicated servers.

    CAUTION: The dedicated Mitsuba server setup uses no authentication

ChangeLog
2010-08-12  Wenzel Jakob <wenzel@cs.cornell.edu>

	* converter.cpp: Much better import of COLLADA scenes from
		blender.

	* importdlg.cpp: User interface for importing scenes.

2010-08-09  Wenzel Jakob <wenzel@cs.cornell.edu>

	* rendersettingsdlg.cpp: Try to maintain the number of 
		samples/pixel when switching between different samplers.

2010-08-05  Wenzel Jakob <wenzel@cs.cornell.edu>
	
	* constant.cpp, src/libhw/vpl.cpp: Support for image environment 
		sources and spot lights in the real-time preview.

2010-08-03  Wenzel Jakob <wenzel@cs.cornell.edu>

	* mitsuba.cpp: Ability to run the 'mitsuba' execubable in arbitrary
		locations on OSX while ensuring that it still finds all plugins etc.

	* mtsimport: COLLADA importer support for ambient lights and 'polylist' meshes.

	* scene.cpp: Create a default camera when none is specified in the XML file.

	* constant.cpp, src/libhw/vpl.cpp: Support for rendering constant environment 
		sources in the real-time preview.

2010-07-21  Wenzel Jakob <wenzel@cs.cornell.edu>

	* glwidget.cpp, mainwindow.cpp: Support for Drag & Drop

	* glwidget.cpp, mainwindow.cpp: The Mitsuba user interface now also
		doubles as an EXR viewer / tonemapper. Drag an EXR file onto the
		UI or open it using the File menu, and the image opens in a new
		tab. Afterwards, it is possible to export the image as a tonemapped
		8-bit PNG image.

2010-07-19  Wenzel Jakob <wenzel@cs.cornell.edu>

	* include/mitsuba/mitsuba.h: First semi-official external release 0.1.1

	* rendersettingsdlg.cpp: Perform a sanity check when changing the rendering
		settings: if the configuration doesn't make any sense, prevent the 
		user from continuing.

	* mainwindow.cpp, many others: Numerous bugfixes for all platforms

2010-07-18  Wenzel Jakob <wenzel@cs.cornell.edu>

	* previewsettingsdlg_cocoa_impl.mm: When running on OSX, use a
		Cocoa-based settings tweaker using a 'HUD Layout' instead of 
		the Qt version.
	
	* debian/*: Finished debian package support
	
	* glwidget.cpp: Fixed some crashes

2010-07-17  Wenzel Jakob <wenzel@cs.cornell.edu>

	* preview.cpp: Many bugfixes for standard ray tracing. The partial
		preview shown while moving through the scene now adapts to its 
		complexity, making the interaction smoother.

	* preview.cpp: Support for coherent ray tracing -- about 2-4x 
		faster than standard ray tracing.

	* glwidget.cpp: When rendering very complex scenes, it was possible
		that the UI required a long time to acquire a image containing
		the most recent preview. This part is now allowed to time out 
		gracefully, leading to a more responsive user interface.

	* mainwindow.cpp: "Resize to Fit" menu action. 

	* mainwindow.cpp: Open/Save dialogs now remember their position,
		layout and directory.

	* debian/*: Support for generating Debian/Ubuntu packages

2010-07-15  Wenzel Jakob <wenzel@cs.cornell.edu>

	* preview.cpp: Very basic support for rendering the preview using 
		standard ray tracing. 

2010-07-14  Wenzel Jakob <wenzel@cs.cornell.edu>

	* errctrl.cpp: Added a feature that enables the error-controlling
		integrator to switch between an image-wide and a per-pixel based
		error criterion.

	* glwidget.cpp, mainwindow.cpp, previewsettingsdlg.cpp: Added support 
		for the tonemapping technique by Reinhard et al. from 2002. It
		dynamically adapts to the average luminance in the viewport.

2010-07-13  Wenzel Jakob <wenzel@cs.cornell.edu>

	* xmltreemodel.cpp: When showing the integrator/sampler/film properties
		in the render setting editor, the less important ones are now drawn
		using a light grey color.

	* triangle.cpp: Fixed a numerical issue when primitive clipping/
		perfect splits were active while building the kd-tree -- in some 
		rare cases, this could result in missing geometry.

	* glwidget.cpp, preview.cpp: Nicer behavior when switching tabs.
		Previously, the preview would regenerate from scratch, which caused
		unpleasant flicker. Now, the preview stores the complete state just
		before switching away, and this is used to resume when switching
		back to the tab.

	* sceneloader.cpp, mainwindow.cpp: The interactive preview now tries
		to detect the path length from the current scene configuration.
	
	* rendersettingsdlg.cpp, save.cpp: It is now possible to configure the
		adaptive integrator & irradiance cache from within the GUI.

2010-07-11  Wenzel Jakob <wenzel@cs.cornell.edu>

	* mainwindow.cpp: Added a 'clone tab' feature, which is accessible
		in the File menu and using a context menu on the tabs

	* many files: Changed the system how scenes are transmitted over the 
		network. The camera and integrator are now serialized separately and 

tools/qt4.py

"""SCons.Tool.qt

Tool-specific initialization for Qt.

There normally shouldn't be any need to import this module directly.
It will usually be imported through the generic SCons.Tool.Tool()
selection method.

"""

#
# Copyright (c) 2001, 2002, 2003, 2004, 2005, 2006, 2007 The SCons Foundation
#
# Permission is hereby granted, free of charge, to any person obtaining
# a copy of this software and associated documentation files (the
# "Software"), to deal in the Software without restriction, including
# without limitation the rights to use, copy, modify, merge, publish,
# distribute, sublicense, and/or sell copies of the Software, and to
# permit persons to whom the Software is furnished to do so, subject to
# the following conditions:
#
# The above copyright notice and this permission notice shall be included
# in all copies or substantial portions of the Software.
#
# THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY
# KIND, EXPRESS OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE
# WARRANTIES OF MERCHANTABILITY, FITNESS FOR A PARTICULAR PURPOSE AND
# NONINFRINGEMENT. IN NO EVENT SHALL THE AUTHORS OR COPYRIGHT HOLDERS BE
# LIABLE FOR ANY CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER IN AN ACTION
# OF CONTRACT, TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN CONNECTION
# WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE SOFTWARE.
#

__revision__ = "/home/scons/scons/branch.0/branch.96/baseline/src/engine/SCons/Tool/qt.py 0.96.92.D001 2006/04/10 23:13:27 knight"

import os.path
import re

import SCons.Action
import SCons.Builder
import SCons.Defaults
import SCons.Scanner
import SCons.Tool
import SCons.Util

class ToolQtWarning(SCons.Warnings.Warning):
	pass

class GeneratedMocFileNotIncluded(ToolQtWarning):
	pass

class QtdirNotFound(ToolQtWarning):
	pass

SCons.Warnings.enableWarningClass(ToolQtWarning)

qrcinclude_re = re.compile(r'<file>([^<]*)</file>', re.M)

def transformToWinePath(path) :
	return os.popen('winepath -w "%s"'%path).read().strip().replace('\\','/')

header_extensions = [".h", ".hxx", ".hpp", ".hh"]
if SCons.Util.case_sensitive_suffixes('.h', '.H'):
	header_extensions.append('.H')
# TODO: The following two lines will work when integrated back to SCons
# TODO: Meanwhile the third line will do the work
#cplusplus = __import__('c++', globals(), locals(), [])
#cxx_suffixes = cplusplus.CXXSuffixes
cxx_suffixes = [".c", ".cxx", ".cpp", ".cc"]

def checkMocIncluded(target, source, env):
	moc = target[0]
	cpp = source[0]
	# looks like cpp.includes is cleared before the build stage :-(
	# not really sure about the path transformations (moc.cwd? cpp.cwd?) :-/
	path = SCons.Defaults.CScan.path_function(env, moc.cwd)
	includes = SCons.Defaults.CScan(cpp, env, path)
	if not moc in includes:
		SCons.Warnings.warn(
			GeneratedMocFileNotIncluded,
			"Generated moc file '%s' is not included by '%s'" %
			(str(moc), str(cpp)))

def find_file(filename, paths, node_factory):
	for dir in paths:
		node = node_factory(filename, dir)
		if node.rexists():
			return node
	return None

class _Automoc:
	"""
	Callable class, which works as an emitter for Programs, SharedLibraries and
	StaticLibraries.
	"""

	def __init__(self, objBuilderName):
		self.objBuilderName = objBuilderName
		
	def __call__(self, target, source, env):
		"""
		Smart autoscan function. Gets the list of objects for the Program
		or Lib. Adds objects and builders for the special qt files.
		"""
		try:
			if int(env.subst('$QT4_AUTOSCAN')) == 0:
				return target, source
		except ValueError:
			pass
		try:
			debug = int(env.subst('$QT4_DEBUG'))
		except ValueError:
			debug = 0
		
		# some shortcuts used in the scanner
		splitext = SCons.Util.splitext
		objBuilder = getattr(env, self.objBuilderName)

		# some regular expressions:
		# Q_OBJECT detection
		q_object_search = re.compile(r'[^A-Za-z0-9]Q_OBJECT[^A-Za-z0-9]') 
		# cxx and c comment 'eater'
		#comment = re.compile(r'(//.*)|(/\*(([^*])|(\*[^/]))*\*/)')
		# CW: something must be wrong with the regexp. See also bug #998222
		#	 CURRENTLY THERE IS NO TEST CASE FOR THAT
		
		# The following is kind of hacky to get builders working properly (FIXME)
		objBuilderEnv = objBuilder.env
		objBuilder.env = env
		mocBuilderEnv = env.Moc4.env
		env.Moc4.env = env
		
		# make a deep copy for the result; MocH objects will be appended
		out_sources = source[:]

		for obj in source:
			if isinstance(obj,basestring):  # big kludge!
				print "scons: qt4: '%s' MAYBE USING AN OLD SCONS VERSION AND NOT CONVERTED TO 'File'. Discarded." % str(obj)
				continue
			if not obj.has_builder():
				# binary obj file provided
				if debug:
					print "scons: qt: '%s' seems to be a binary. Discarded." % str(obj)
				continue
			cpp = obj.sources[0]
			if not splitext(str(cpp))[1] in cxx_suffixes:
				if debug:
					print "scons: qt: '%s' is no cxx file. Discarded." % str(cpp) 
				# c or fortran source
				continue
			#cpp_contents = comment.sub('', cpp.get_contents())
			try:
				cpp_contents = cpp.get_contents()
			except: continue # may be an still not generated source
			h=None
			for h_ext in header_extensions:
				# try to find the header file in the corresponding source
				# directory
				hname = splitext(cpp.name)[0] + h_ext
				h = find_file(hname, (cpp.get_dir(),), env.File)
				if h:
					if debug:
						print "scons: qt: Scanning '%s' (header of '%s')" % (str(h), str(cpp))
					#h_contents = comment.sub('', h.get_contents())
					h_contents = h.get_contents()
					break
			if not h and debug:
				print "scons: qt: no header for '%s'." % (str(cpp))
			if h and q_object_search.search(h_contents):
				# h file with the Q_OBJECT macro found -> add moc_cpp
				moc_cpp = env.Moc4(h)
				moc_o = objBuilder(moc_cpp)
				out_sources.append(moc_o)
				#moc_cpp.target_scanner = SCons.Defaults.CScan
				if debug:
					print "scons: qt: found Q_OBJECT macro in '%s', moc'ing to '%s'" % (str(h), str(moc_cpp))
			if cpp and q_object_search.search(cpp_contents):
				# cpp file with Q_OBJECT macro found -> add moc
				# (to be included in cpp)
				moc = env.Moc4(cpp)
				env.Ignore(moc, moc)
				if debug:
					print "scons: qt: found Q_OBJECT macro in '%s', moc'ing to '%s'" % (str(cpp), str(moc))
				#moc.source_scanner = SCons.Defaults.CScan
		# restore the original env attributes (FIXME)
		objBuilder.env = objBuilderEnv
		env.Moc4.env = mocBuilderEnv

		return (target, out_sources)

AutomocShared = _Automoc('SharedObject')
AutomocStatic = _Automoc('StaticObject')

def _detect(env):
	"""Not really safe, but fast method to detect the QT library"""
	try: return env['QTDIR']
	except KeyError: pass

	try: return os.environ['QTDIR']
	except KeyError: pass

	moc = env.WhereIs('moc-qt4') or env.WhereIs('moc4') or env.WhereIs('moc')
	if moc:
		QTDIR = os.path.dirname(os.path.dirname(moc))
		SCons.Warnings.warn(
			QtdirNotFound,
			"QTDIR variable is not defined, using moc executable as a hint (QTDIR=%s)" % QTDIR)
		return QTDIR

	raise SCons.Errors.StopError(
		QtdirNotFound,
		"Could not detect Qt 4 installation")
	return None

def generate(env):
	"""Add Builders and construction variables for qt to an Environment."""

	def locateQt4Command(env, command, qtdir) :
		suffixes = [
			'-qt4',
			'-qt4.exe',
			'4',
			'4.exe',
			'',
			'.exe',
		]
		triedPaths = []
		for suffix in suffixes :
			fullpath = os.path.join(qtdir,'bin',command + suffix)
			if os.access(fullpath, os.X_OK) :
				return fullpath
			triedPaths.append(fullpath)

		fullpath = env.Detect([command+'-qt4', command+'4', command])
		if not (fullpath is None) : return fullpath

		raise Exception("Qt4 command '" + command + "' not found. Tried: " + ', '.join(triedPaths))
		

	CLVar = SCons.Util.CLVar
	Action = SCons.Action.Action
	Builder = SCons.Builder.Builder
	splitext = SCons.Util.splitext

	env['QTDIR']  = _detect(env)
	# TODO: 'Replace' should be 'SetDefault'
#	env.SetDefault(
	env.Replace(
		QTDIR  = _detect(env),
		QT4_BINPATH = os.path.join('$QTDIR', 'bin'),
		# TODO: This is not reliable to QTDIR value changes but needed in order to support '-qt4' variants
		QT4_MOC = locateQt4Command(env,'moc', env['QTDIR']),
		QT4_UIC = locateQt4Command(env,'uic', env['QTDIR']),
		QT4_RCC = locateQt4Command(env,'rcc', env['QTDIR']),
		QT4_LUPDATE = locateQt4Command(env,'lupdate', env['QTDIR']),
		QT4_LRELEASE = locateQt4Command(env,'lrelease', env['QTDIR']),

		QT4_AUTOSCAN = 1, # Should the qt tool try to figure out, which sources are to be moc'ed?

		# Some QT specific flags. I don't expect someone wants to
		# manipulate those ...
		QT4_UICFLAGS = CLVar(''),
		QT4_MOCFROMHFLAGS = CLVar(''),
		QT4_MOCFROMCXXFLAGS = CLVar('-i'),
		QT4_QRCFLAGS = '',

		# suffixes/prefixes for the headers / sources to generate
		QT4_UISUFFIX = '.ui',
		QT4_UICDECLPREFIX = 'ui_',
		QT4_UICDECLSUFFIX = '.h',
		QT4_MOCINCPREFIX = '-I',
		QT4_MOCHPREFIX = 'moc_',
		QT4_MOCHSUFFIX = '$CXXFILESUFFIX',
		QT4_MOCCXXPREFIX = '',
		QT4_MOCCXXSUFFIX = '.moc',
		QT4_QRCSUFFIX = '.qrc',
		QT4_QRCCXXSUFFIX = '$CXXFILESUFFIX',
		QT4_QRCCXXPREFIX = 'qrc_',
		QT4_MOCCPPPATH = [],
		QT4_MOCINCFLAGS = '$( ${_concat(QT4_MOCINCPREFIX, QT4_MOCCPPPATH, INCSUFFIX, __env__, RDirs)} $)',

		# Commands for the qt support ...
		QT4_UICCOM = '$QT4_UIC $QT4_UICFLAGS -o $TARGET $SOURCE',
		QT4_MOCFROMHCOM = '$QT4_MOC $QT4_MOCFROMHFLAGS $QT4_MOCINCFLAGS -o $TARGET $SOURCE',
		QT4_MOCFROMCXXCOM = [
			'$QT4_MOC $QT4_MOCFROMCXXFLAGS $QT4_MOCINCFLAGS -o $TARGET $SOURCE',
			Action(checkMocIncluded,None)],
		QT4_LUPDATECOM = '$QT4_LUPDATE $SOURCE -ts $TARGET',
		QT4_LRELEASECOM = '$QT4_LRELEASE $SOURCE',
		QT4_RCCCOM = '$QT4_RCC $QT4_QRCFLAGS $SOURCE -o $TARGET',
		)

	# Translation builder
	tsbuilder = Builder(
		action = SCons.Action.Action('$QT4_LUPDATECOM'), #,'$QT4_LUPDATECOMSTR'),
		multi=1
		)
	env.Append( BUILDERS = { 'Ts': tsbuilder } )
	qmbuilder = Builder(
		action = SCons.Action.Action('$QT4_LRELEASECOM'),# , '$QT4_LRELEASECOMSTR'),
		src_suffix = '.ts',
		suffix = '.qm',
		single_source = True
		)
	env.Append( BUILDERS = { 'Qm': qmbuilder } )

	# Resource builder
	def scanResources(node, env, path, arg):
		# I've being careful on providing names relative to the qrc file
		# If that was not needed that code could be simplified a lot
		def recursiveFiles(basepath, path) :
			result = []
			for item in os.listdir(os.path.join(basepath, path)) :
				itemPath = os.path.join(path, item)
				if os.path.isdir(os.path.join(basepath, itemPath)) :
					result += recursiveFiles(basepath, itemPath)
				else:
					result.append(itemPath)
			return result
		contents = node.get_contents()
		includes = qrcinclude_re.findall(contents)
		qrcpath = os.path.dirname(node.path)
		dirs = [included for included in includes if os.path.isdir(os.path.join(qrcpath,included))]
		# dirs need to include files recursively
		for dir in dirs :
			includes.remove(dir)
			includes+=recursiveFiles(qrcpath,dir)
		return includes
	qrcscanner = SCons.Scanner.Scanner(name = 'qrcfile',
		function = scanResources,
		argument = None,
		skeys = ['.qrc'])
	qrcbuilder = Builder(
		action = SCons.Action.Action('$QT4_RCCCOM', '$QT4_RCCCOMSTR'),
		source_scanner = qrcscanner,
		src_suffix = '$QT4_QRCSUFFIX',
		suffix = '$QT4_QRCCXXSUFFIX',
		prefix = '$QT4_QRCCXXPREFIX',
		single_source = True
		)
	env.Append( BUILDERS = { 'Qrc': qrcbuilder } )

	# Interface builder
	uic4builder = Builder(
		action = SCons.Action.Action('$QT4_UICCOM', '$QT4_UICCOMSTR'),
		src_suffix='$QT4_UISUFFIX',
		suffix='$QT4_UICDECLSUFFIX',
		prefix='$QT4_UICDECLPREFIX',
		single_source = True
		#TODO: Consider the uiscanner on new scons version
		)
	env['BUILDERS']['Uic4'] = uic4builder

	# Metaobject builder
	mocBld = Builder(action={}, prefix={}, suffix={})
	for h in header_extensions:
		act = SCons.Action.Action('$QT4_MOCFROMHCOM', '$QT4_MOCFROMHCOMSTR')
		mocBld.add_action(h, act)
		mocBld.prefix[h] = '$QT4_MOCHPREFIX'
		mocBld.suffix[h] = '$QT4_MOCHSUFFIX'
	for cxx in cxx_suffixes:
		act = SCons.Action.Action('$QT4_MOCFROMCXXCOM', '$QT4_MOCFROMCXXCOMSTR')
		mocBld.add_action(cxx, act)
		mocBld.prefix[cxx] = '$QT4_MOCCXXPREFIX'
		mocBld.suffix[cxx] = '$QT4_MOCCXXSUFFIX'
	env['BUILDERS']['Moc4'] = mocBld

	# er... no idea what that was for
	static_obj, shared_obj = SCons.Tool.createObjBuilders(env)
	static_obj.src_builder.append('Uic4')
	shared_obj.src_builder.append('Uic4')

	# We use the emitters of Program / StaticLibrary / SharedLibrary
	# to scan for moc'able files
	# We can't refer to the builders directly, we have to fetch them
	# as Environment attributes because that sets them up to be called
	# correctly later by our emitter.
	env.AppendUnique(PROGEMITTER =[AutomocStatic],
					 SHLIBEMITTER=[AutomocShared],
					 LIBEMITTER  =[AutomocStatic],
					)

	# TODO: Does dbusxml2cpp need an adapter
	env.AddMethod(enable_modules, "EnableQt4Modules")

def enable_modules(self, modules, debug=False, crosscompiling=False) :
	import sys

	validModules = [
		'QtCore',
		'QtGui',
		'QtOpenGL',
		'Qt3Support',
		'QtAssistant', # deprecated
		'QtAssistantClient',
		'QtScript',
		'QtDBus',
		'QtSql',
		'QtSvg',
		# The next modules have not been tested yet so, please
		# maybe they require additional work on non Linux platforms
		'QtNetwork',
		'QtTest',
		'QtXml',
		'QtXmlPatterns',
		'QtUiTools',
		'QtDesigner',
		'QtDesignerComponents',
		'QtWebKit',
		'QtHelp',
		'QtScript',
		'QtScriptTools',
		'QtMultimedia',
		]
	pclessModules = [
# in qt <= 4.3 designer and designerComponents are pcless, on qt4.4 they are not, so removed.	
#		'QtDesigner',
#		'QtDesignerComponents',
	]
	staticModules = [
		'QtUiTools',
	]
	invalidModules=[]
	for module in modules:
		if module not in validModules :
			invalidModules.append(module)
	if invalidModules :
		raise Exception("Modules %s are not Qt4 modules. Valid Qt4 modules are: %s"% (
			str(invalidModules),str(validModules)))

	moduleDefines = {
		'QtScript'   : ['QT_SCRIPT_LIB'],
		'QtSvg'      : ['QT_SVG_LIB'],
		'Qt3Support' : ['QT_QT3SUPPORT_LIB','QT3_SUPPORT'],
		'QtSql'      : ['QT_SQL_LIB'],
		'QtXml'      : ['QT_XML_LIB'],
		'QtOpenGL'   : ['QT_OPENGL_LIB'],
		'QtGui'      : ['QT_GUI_LIB'],
		'QtNetwork'  : ['QT_NETWORK_LIB'],
		'QtCore'     : ['QT_CORE_LIB'],
	}
	for module in modules :
		try : self.AppendUnique(CPPDEFINES=moduleDefines[module])
		except: pass
	debugSuffix = ''
	if sys.platform in ["darwin", "linux2"] and not crosscompiling :
		if debug : debugSuffix = '_debug'
		for module in modules :
			if module not in pclessModules : continue
			self.AppendUnique(LIBS=[module+debugSuffix])
			self.AppendUnique(LIBPATH=[os.path.join("$QTDIR","lib")])
			self.AppendUnique(CPPPATH=[os.path.join("$QTDIR","include","qt4")])
			self.AppendUnique(CPPPATH=[os.path.join("$QTDIR","include","qt4",module)])
		pcmodules = [module+debugSuffix for module in modules if module not in pclessModules ]
		if 'QtDBus' in pcmodules:
			self.AppendUnique(CPPPATH=[os.path.join("$QTDIR","include","qt4","QtDBus")])
		if "QtAssistant" in pcmodules:
			self.AppendUnique(CPPPATH=[os.path.join("$QTDIR","include","qt4","QtAssistant")])
			pcmodules.remove("QtAssistant")
			pcmodules.append("QtAssistantClient")
		if sys.platform == "linux2":
			self.ParseConfig('pkg-config %s --libs --cflags'% ' '.join(pcmodules))
		else:
			for module in pcmodules:
				self.AppendUnique(CPPPATH="/Library/Frameworks/%s.framework/Versions/4/Headers" % module)
				self.Append(LINKFLAGS=['-framework', module])
		self["QT4_MOCCPPPATH"] = self["CPPPATH"]
		return
	if sys.platform == "win32" or crosscompiling :
		if crosscompiling:
			transformedQtdir = transformToWinePath(self['QTDIR'])
			self['QT4_MOC'] = "QTDIR=%s %s"%( transformedQtdir, self['QT4_MOC'])
		self.AppendUnique(CPPPATH=[os.path.join("$QTDIR","include")])
		try: modules.remove("QtDBus")
		except: pass
		if debug : debugSuffix = 'd'
		if "QtAssistant" in modules:
			self.AppendUnique(CPPPATH=[os.path.join("$QTDIR","include","QtAssistant")])
			modules.remove("QtAssistant")
			modules.append("QtAssistantClient")
		self.AppendUnique(LIBS=[lib+'4'+debugSuffix for lib in modules if lib not in staticModules])
		self.PrependUnique(LIBS=[lib+debugSuffix for lib in modules if lib in staticModules])
		if 'QtOpenGL' in modules:
			self.AppendUnique(LIBS=['opengl32'])
		self.AppendUnique(CPPPATH=[ '$QTDIR/include/'])
		self.AppendUnique(CPPPATH=[ '$QTDIR/include/'+module for module in modules])
		if crosscompiling :
			self["QT4_MOCCPPPATH"] = [
				path.replace('$QTDIR', transformedQtdir)
					for path in self['CPPPATH'] ]
		else :
			self["QT4_MOCCPPPATH"] = self["CPPPATH"]
		self.AppendUnique(LIBPATH=[os.path.join('$QTDIR','lib')])
		return
	"""
	if sys.platform=="darwin" :
		# TODO: Test debug version on Mac
		self.AppendUnique(LIBPATH=[os.path.join('$QTDIR','lib')])
		self.AppendUnique(LINKFLAGS="-F$QTDIR/lib")
		self.AppendUnique(LINKFLAGS="-L$QTDIR/lib") #TODO clean!
		if debug : debugSuffix = 'd'
		for module in modules :
#			self.AppendUnique(CPPPATH=[os.path.join("$QTDIR","include")])
#			self.AppendUnique(CPPPATH=[os.path.join("$QTDIR","include",module)])
# port qt4-mac:
			self.AppendUnique(CPPPATH=[os.path.join("$QTDIR","include", "qt4")])
			self.AppendUnique(CPPPATH=[os.path.join("$QTDIR","include", "qt4", module)])
			if module in staticModules :
				self.AppendUnique(LIBS=[module+debugSuffix]) # TODO: Add the debug suffix
				self.AppendUnique(LIBPATH=[os.path.join("$QTDIR","lib")])
			else :
#				self.Append(LINKFLAGS=['-framework', module])
# port qt4-mac:
				self.Append(LIBS=module)
		if 'QtOpenGL' in modules:
			self.AppendUnique(LINKFLAGS="-F/System/Library/Frameworks")
			self.Append(LINKFLAGS=['-framework', 'AGL']) #TODO ughly kludge to avoid quotes
			self.Append(LINKFLAGS=['-framework', 'OpenGL'])
		self["QT4_MOCCPPPATH"] = self["CPPPATH"]
		return
# This should work for mac but doesn't
#	env.AppendUnique(FRAMEWORKPATH=[os.path.join(env['QTDIR'],'lib')])
#	env.AppendUnique(FRAMEWORKS=['QtCore','QtGui','QtOpenGL', 'AGL'])
	"""


def exists(env):
	return _detect(env)
